packages/drivers/odsp-driver/src/odspDocumentService.ts

```diff
@@ -304,9 +297,67 @@ export class OdspDocumentService implements IDocumentService {
         });
     }
 
+    private clearJoinSessionTimer() {
+        if (this.joinSessionRefreshTimer !== undefined) {
+            clearTimeout(this.joinSessionRefreshTimer);
+            this.joinSessionRefreshTimer = undefined;
+        }
+    }
+
+    private async refreshSessionPeriodically(requestWebsocketTokenFromJoinSession: boolean) {
+        for (let i = 0;;i++) {
+            // If the refresh time is not defined, then let the session expire.
+            if (this.joinSessionRefreshTime === undefined || this.joinSessionRefreshTime <= 0) {
+                this.mc.logger.sendErrorEvent({
+                    eventName: "JoinSessionRefreshTimeNotSet",
+                    joinSessionRefreshTime: this.joinSessionRefreshTime,
+                    timerRunning: this.joinSessionRefreshTimer !== undefined,
+                });
+                break;
+            }
+            await new Promise<void>((resolve) => {
+                assert(this.joinSessionRefreshTime !== undefined, "Join session Refresh time should be defined");
+                this.joinSessionRefreshTimer = setTimeout(async () => {
```

Comment:
setTimeout() probably does not care about return type, so async does nothing here.
Which begs question - how do we report errors / exceptions? Sounds like with current code they become unhandled exceptions, which is not great. Prefer to catch all errors and report them to our telemetry.

Reply:
Async should be used as we are awaiting inside it for the network call to be completed. Also catching the error in the caller and firing the error event with appropriate props.